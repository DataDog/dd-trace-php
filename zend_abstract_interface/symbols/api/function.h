#ifndef HAVE_ZAI_SYMBOLS_API_FUNCTION_H
#define HAVE_ZAI_SYMBOLS_API_FUNCTION_H

static inline
zend_function *zai_symbol_lookup_function(zai_symbol_scope_t scope_type,
                                          void *scope,
                                          zai_str *name) {
    return (zend_function *)zai_symbol_lookup(ZAI_SYMBOL_TYPE_FUNCTION, scope_type, scope, name	);
}

static inline zend_function *zai_symbol_lookup_function_global(zai_str vfn) {
    return zai_symbol_lookup_function(ZAI_SYMBOL_SCOPE_GLOBAL, NULL, &vfn);
}

static inline
zend_function *zai_symbol_lookup_function_ns(zai_str vns, zai_str vfn) {
    return zai_symbol_lookup_function(ZAI_SYMBOL_SCOPE_NAMESPACE, &vns, &vfn);
}

#endif
