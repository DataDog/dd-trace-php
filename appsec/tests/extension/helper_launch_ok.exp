helper_mgr_acquire_conn run
bool(true)

Connected?
bool(true)

Contents of helper log:
[%s] pre-exec: Going for second fork
[%s] pre-exec: Intermediate process exiting
[%s] pre-exec: About to call execv
Checking open file descriptors
* has file descriptor 0
* has file descriptor 1
* has file descriptor 2
Checking procmask
array (
)
Checking umask
0
Checking parent uid (should be 1)
1
Checking process group id == pid (is a process group leader)
OK
Checking session id == pid (is a session leader)
OK

Checking socket id
file descriptor is %d
opened fd %d

Accepting a connection:
accepted a new connection
read initial message from extension (size %d)
read remaining data