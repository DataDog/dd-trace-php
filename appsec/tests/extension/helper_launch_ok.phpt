--TEST--
Test that the helper is launched with the correct properties
--SKIPIF--
<?php
include __DIR__ . '/inc/check_extension.php';
if (key_exists('USE_ZEND_ALLOC', $_ENV) && $_ENV['USE_ZEND_ALLOC'] == '0') {
    die('skip not to be run with valgrind (does not get adopted by init)');
}
check_extension('posix');
check_extension('pcntl');
check_extension('sockets');
?>
--INI--
datadog.appsec.helper_log_file=/tmp/helper_test.log
datadog.appsec.helper_launch=1
datadog.appsec.log_file=/tmp/php_appsec_test.log
datadog.appsec.log_level=debug
--FILE--
<?php
use function datadog\appsec\testing\{set_helper_path,set_helper_extra_args,is_connected_to_helper,helper_mgr_acquire_conn};

if (key_exists('TEST_PHP_EXECUTABLE', $_ENV)) {
    set_helper_path($_ENV['TEST_PHP_EXECUTABLE']);
} else {
    set_helper_path(readlink('/proc/self/exe'));
}
$ext_dir = ini_get('extension_dir');
$ext_args = array();
foreach(array('posix', 'pcntl', 'sockets') as $ext) {
    if (file_exists("$ext_dir/$ext.so")) {
        $ext_args[] = '-d';
        $ext_args[] = "extension=$ext.so";
    }
}
$helper_args = implode(' ', array_merge(
    array(
        '-n',
        '-d', 'variables_order=EGPCS',
        '-d', 'extension_dir=' . $ext_dir
    ),
    $ext_args,
    array(
        '-d', 'display_errors=1',
        '-d', 'error_reporting=2147483647',
        __DIR__ . "/inc/helper_invocation.php",
        phpversion('ddappsec')
    )));
set_helper_extra_args($helper_args);

echo 'helper_mgr_acquire_conn run', "\n";
var_dump(helper_mgr_acquire_conn());
echo "\n";

echo 'Connected?', "\n";
var_dump(is_connected_to_helper());
echo "\n";

echo 'Contents of helper log:', "\n";
echo file_get_contents('/tmp/helper_test.log');
?>
--EXPECTF--
helper_mgr_acquire_conn run
bool(true)

Connected?
bool(true)

Contents of helper log:
[%s] pre-exec: Going for second fork
[%s] pre-exec: Intermediate process exiting
[%s] pre-exec: About to call execv
Checking open file descriptors
* has file descriptor 0
* has file descriptor 1
* has file descriptor 2
Checking procmask
array (
)
Checking umask
0
Checking parent uid (should be 1)
1
Checking process group id == pid (is a process group leader)
OK
Checking session id == pid (is a session leader)
OK

Checking socket id
file descriptor is %d
opened fd %d

Accepting a connection:
accepted a new connection
read initial message from extension (size %d)
read remaining data
