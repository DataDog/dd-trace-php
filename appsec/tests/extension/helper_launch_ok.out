helper_mgr_acquire_conn run
bool(true)

Connected?
bool(true)

Contents of helper log:
[2023-09-12 14:43:44.324] pre-exec: Going for second fork
[2023-09-12 14:43:44.324] pre-exec: Intermediate process exiting
[2023-09-12 14:43:44.324] pre-exec: About to call execv
Checking open file descriptors
* has file descriptor 0
* has file descriptor 1
* has file descriptor 2
Checking procmask
array (
)
Checking umask
0
Checking parent uid (should be 1)
4770
Checking process group id == pid (is a process group leader)
OK
Checking session id == pid (is a session leader)
OK

Checking socket id
file descriptor is 8
opened fd 8

Accepting a connection:
accepted a new connection
read initial message from extension (size 871)
read remaining data